src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -428,11 +444,68 @@ private static double getRangesTotalSize(List<Splitter.WeightedRange> ranges)
             spaceAvailable -= compaction.totSizeInBytes;
         }
 
+        return new CompactionLimits(runningCompactions,
+                                    maxCompactions,
+                                    maxConcurrentCompactions,
+                                    perLevel,
+                                    levelCount,
+                                    spaceAvailable,
+                                    rateLimitLog);
+    }
+    
+    private Collection<CompactionAggregate> getNextCompactionAggregates(int gcBefore)
+    {
+        final CompactionLimits limits = getCurrentLimits(controller.maxConcurrentCompactions());
         logger.debug("Selecting up to {} new compactions of up to {}, concurrency limit {}{}",
-                     maxCompactions - runningCompactions,
-                     FBUtilities.prettyPrintMemory(spaceAvailable),
-                     maxConcurrentCompactions,
-                     rateLimitLog);
+                     Math.max(0, limits.maxCompactions - limits.runningCompactions),
+                     FBUtilities.prettyPrintMemory(limits.spaceAvailable),
+                     limits.maxConcurrentCompactions,
+                     limits.rateLimitLog);
+
+        maybeUpdateSelector();
+
+        List<CompactionAggregate.UnifiedAggregate> pending = getPendingCompactionAggregates(limits.spaceAvailable, gcBefore);
+
+        for (CompactionAggregate.UnifiedAggregate aggregate : pending)
+        {
+            // The space overhead limit also applies when a single compaction is above that limit. This should
+            // prevent running out of space at the expense of several highest-level tables extra, i.e. slightly
+            // higher read amplification, which I think is a sensible tradeoff; however, operators must be warned
+            // if this happens.
+            warnIfSizeAbove(aggregate, limits.spaceAvailable);
+
+            // Make sure the level count includes all levels for which we have sstables (to be ready to compact
+            // as soon as the threshold is crossed)...
+            limits.levelCount = Math.max(limits.levelCount, aggregate.bucketIndex() + 1);
+            if (aggregate.selected != null)
```

Comment:
Isn't it equivalent to?
```suggestion
            if (aggregate.getSelected() != null)
```